docs/man/man3/QKlipperMessage.3
.TH "QKlipperMessage" 3 "Version 0.2" "QKlipper" \" -*- nroff -*-
.ad l
.nh
.SH NAME
QKlipperMessage
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <qklippermessage\&.h>\fP
.PP
Inherits QObject\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBProtocol\fP { \fBRpcProtocol\fP = 1, \fBHttpProtocol\fP = 2 }"
.br
.ti -1c
.RI "enum \fBState\fP { \fBNone\fP = 0, \fBSending\fP = 1, \fBComplete\fP = 2, \fBError\fP = 3, \fBTimeout\fP = 4 }"
.br
.ti -1c
.RI "enum \fBOrigin\fP { \fBSystem\fP = 1, \fBUser\fP = 2 }"
.br
.in -1c
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBsetId\fP (qint32 \fBid\fP)"
.br
.ti -1c
.RI "void \fBresetId\fP ()"
.br
.ti -1c
.RI "void \fBsetProtocol\fP (\fBProtocol\fP \fBprotocol\fP)"
.br
.ti -1c
.RI "void \fBresetProtocol\fP ()"
.br
.ti -1c
.RI "void \fBsetParams\fP (const QMap< QString, QVariant > &\fBparams\fP)"
.br
.ti -1c
.RI "void \fBsetParam\fP (QString key, QVariant value)"
.br
.ti -1c
.RI "void \fBresetParams\fP ()"
.br
.ti -1c
.RI "void \fBsetMethod\fP (const QString &\fBmethod\fP)"
.br
.ti -1c
.RI "void \fBresetMethod\fP ()"
.br
.ti -1c
.RI "void \fBsetState\fP (\fBState\fP \fBstate\fP)"
.br
.ti -1c
.RI "void \fBresetState\fP ()"
.br
.ti -1c
.RI "void \fBsetResponse\fP (const QJsonValue &\fBresponse\fP)"
.br
.ti -1c
.RI "void \fBsetResponse\fP (const QByteArray &\fBresponse\fP)"
.br
.ti -1c
.RI "void \fBresetResponse\fP ()"
.br
.ti -1c
.RI "void \fBsetErrorString\fP (const QString &\fBerrorString\fP)"
.br
.ti -1c
.RI "void \fBresetErrorString\fP ()"
.br
.ti -1c
.RI "void \fBstartTimer\fP ()"
.br
.ti -1c
.RI "void \fBsetOrigin\fP (\fBOrigin\fP \fBorigin\fP)"
.br
.ti -1c
.RI "void \fBresetOrigin\fP ()"
.br
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBprotocolChanged\fP ()"
.br
.ti -1c
.RI "void \fBmethodChanged\fP ()"
.br
.ti -1c
.RI "void \fBparamsChanged\fP ()"
.br
.ti -1c
.RI "void \fBidChanged\fP ()"
.br
.ti -1c
.RI "void \fBstateChanged\fP ()"
.br
.ti -1c
.RI "void \fBresponseChanged\fP ()"
.br
.ti -1c
.RI "void \fBerrorStringChanged\fP ()"
.br
.ti -1c
.RI "void \fBresponseTimeout\fP ()"
.br
.ti -1c
.RI "void \fBoriginChanged\fP ()"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBQKlipperMessage\fP (QObject *parent=nullptr)"
.br
.ti -1c
.RI "\fB~QKlipperMessage\fP ()"
.br
.ti -1c
.RI "qint32 \fBid\fP () const"
.br
.ti -1c
.RI "\fBProtocol\fP \fBprotocol\fP () const"
.br
.ti -1c
.RI "QMap< QString, QVariant > \fBparams\fP () const"
.br
.ti -1c
.RI "QVariant \fBparam\fP (QString key) const"
.br
.ti -1c
.RI "QString \fBmethod\fP () const"
.br
.ti -1c
.RI "QString \fBtoUri\fP (bool parameterEncoded=true) const"
.br
.ti -1c
.RI "QByteArray \fBtoRpc\fP () const"
.br
.ti -1c
.RI "\fBState\fP \fBstate\fP () const"
.br
.ti -1c
.RI "QJsonValue \fBresponse\fP () const"
.br
.ti -1c
.RI "QString \fBerrorString\fP () const"
.br
.ti -1c
.RI "\fBOrigin\fP \fBorigin\fP () const"
.br
.in -1c
.SS "Protected Slots"

.in +1c
.ti -1c
.RI "void \fBon_responseTimerTimeout\fP ()"
.br
.in -1c
.SS "Properties"

.in +1c
.ti -1c
.RI "qint32 \fBid\fP"
.br
.ti -1c
.RI "\fBProtocol\fP \fBprotocol\fP"
.br
.ti -1c
.RI "QMap< QString, QVariant > \fBparams\fP"
.br
.ti -1c
.RI "QString \fBmethod\fP"
.br
.ti -1c
.RI "\fBState\fP \fBstate\fP"
.br
.ti -1c
.RI "QJsonValue \fBresponse\fP"
.br
.ti -1c
.RI "QString \fBerrorString\fP"
.br
.ti -1c
.RI "\fBOrigin\fP \fBorigin\fP"
.br
.in -1c
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBQKlipperMessage::Origin\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fISystem \fP\fP
.TP
\fB\fIUser \fP\fP
.SS "enum \fBQKlipperMessage::Protocol\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIRpcProtocol \fP\fP
.TP
\fB\fIHttpProtocol \fP\fP
.SS "enum \fBQKlipperMessage::State\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fINone \fP\fP
.TP
\fB\fISending \fP\fP
.TP
\fB\fIComplete \fP\fP
.TP
\fB\fIError \fP\fP
.TP
\fB\fITimeout \fP\fP
.SH "Constructor & Destructor Documentation"
.PP 
.SS "QKlipperMessage::QKlipperMessage (QObject * parent = \fCnullptr\fP)\fC [explicit]\fP"

.SS "QKlipperMessage::~QKlipperMessage ()"

.SH "Member Function Documentation"
.PP 
.SS "QString QKlipperMessage::errorString () const"

.SS "void QKlipperMessage::errorStringChanged ()\fC [signal]\fP"

.SS "qint32 QKlipperMessage::id () const"

.SS "void QKlipperMessage::idChanged ()\fC [signal]\fP"

.SS "QString QKlipperMessage::method () const"

.SS "void QKlipperMessage::methodChanged ()\fC [signal]\fP"

.SS "void QKlipperMessage::on_responseTimerTimeout ()\fC [protected]\fP, \fC [slot]\fP"

.SS "\fBQKlipperMessage::Origin\fP QKlipperMessage::origin () const"

.SS "void QKlipperMessage::originChanged ()\fC [signal]\fP"

.SS "QVariant QKlipperMessage::param (QString key) const"

.SS "QMap< QString, QVariant > QKlipperMessage::params () const"

.SS "void QKlipperMessage::paramsChanged ()\fC [signal]\fP"

.SS "\fBQKlipperMessage::Protocol\fP QKlipperMessage::protocol () const"

.SS "void QKlipperMessage::protocolChanged ()\fC [signal]\fP"

.SS "void QKlipperMessage::resetErrorString ()\fC [slot]\fP"

.SS "void QKlipperMessage::resetId ()\fC [slot]\fP"

.SS "void QKlipperMessage::resetMethod ()\fC [slot]\fP"

.SS "void QKlipperMessage::resetOrigin ()\fC [slot]\fP"

.SS "void QKlipperMessage::resetParams ()\fC [slot]\fP"

.SS "void QKlipperMessage::resetProtocol ()\fC [slot]\fP"

.SS "void QKlipperMessage::resetResponse ()\fC [slot]\fP"

.SS "void QKlipperMessage::resetState ()\fC [slot]\fP"

.SS "QJsonValue QKlipperMessage::response () const"

.SS "void QKlipperMessage::responseChanged ()\fC [signal]\fP"

.SS "void QKlipperMessage::responseTimeout ()\fC [signal]\fP"

.SS "void QKlipperMessage::setErrorString (const QString & errorString)\fC [slot]\fP"

.SS "void QKlipperMessage::setId (qint32 id)\fC [slot]\fP"

.SS "void QKlipperMessage::setMethod (const QString & method)\fC [slot]\fP"

.SS "void QKlipperMessage::setOrigin (\fBOrigin\fP origin)\fC [slot]\fP"

.SS "void QKlipperMessage::setParam (QString key, QVariant value)\fC [slot]\fP"

.SS "void QKlipperMessage::setParams (const QMap< QString, QVariant > & params)\fC [slot]\fP"

.SS "void QKlipperMessage::setProtocol (\fBProtocol\fP protocol)\fC [slot]\fP"

.SS "void QKlipperMessage::setResponse (const QByteArray & response)\fC [slot]\fP"

.SS "void QKlipperMessage::setResponse (const QJsonValue & response)\fC [slot]\fP"

.SS "void QKlipperMessage::setState (\fBState\fP state)\fC [slot]\fP"

.SS "void QKlipperMessage::startTimer ()\fC [slot]\fP"

.SS "\fBQKlipperMessage::State\fP QKlipperMessage::state () const"

.SS "void QKlipperMessage::stateChanged ()\fC [signal]\fP"

.SS "QByteArray QKlipperMessage::toRpc () const"

.SS "QString QKlipperMessage::toUri (bool parameterEncoded = \fCtrue\fP) const"

.SH "Property Documentation"
.PP 
.SS "QString QKlipperMessage::errorString\fC [read]\fP, \fC [write]\fP"

.SS "qint32 QKlipperMessage::id\fC [read]\fP, \fC [write]\fP"

.SS "QString QKlipperMessage::method\fC [read]\fP, \fC [write]\fP"

.SS "\fBOrigin\fP QKlipperMessage::origin\fC [read]\fP, \fC [write]\fP"

.SS "QMap<QString, QVariant> QKlipperMessage::params\fC [read]\fP, \fC [write]\fP"

.SS "\fBProtocol\fP QKlipperMessage::protocol\fC [read]\fP, \fC [write]\fP"

.SS "QJsonValue QKlipperMessage::response\fC [read]\fP, \fC [write]\fP"

.SS "\fBState\fP QKlipperMessage::state\fC [read]\fP, \fC [write]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for QKlipper from the source code\&.


QKlipperMessage/qklippermessage.cpp
#include "qklippermessage.h"

qint32 QKlipperMessage::s_currentId = 0;

QKlipperMessage::QKlipperMessage(QObject *parent)
    : QObject{parent}
{
    m_id = ++s_currentId;

    m_responseTimer = new QTimer();
    m_responseTimer->setInterval(3000);
    m_responseTimer->setSingleShot(true);

    connect(m_responseTimer, SIGNAL(timeout()), this, SLOT(on_responseTimerTimeout()));
}

QKlipperMessage::~QKlipperMessage()
{
    if(m_responseTimer)
        delete m_responseTimer;
}

qint32 QKlipperMessage::id() const
{
    return m_id;
}

void QKlipperMessage::setId(qint32 id)
{
    if (m_id == id)
        return;

    m_id = id;
    emit idChanged();
}

void QKlipperMessage::resetId()
{
    setId(0);
}

QKlipperMessage::Protocol QKlipperMessage::protocol() const
{
    return m_protocol;
}

void QKlipperMessage::setProtocol(Protocol protocol)
{
    if (m_protocol == protocol)
        return;

    m_protocol = protocol;
    emit protocolChanged();
}

void QKlipperMessage::resetProtocol()
{
    setProtocol(HttpProtocol);
}

QMap<QString, QVariant> QKlipperMessage::params() const
{
    return m_params;
}

QVariant QKlipperMessage::param(QString key) const
{
    if(m_params.contains(key))
        return m_params[key];

    return QVariant();
}

void QKlipperMessage::setParams(const QMap<QString, QVariant> &params)
{
    if (m_params == params)
        return;

    m_params = params;
    emit paramsChanged();
}

void QKlipperMessage::setParam(QString key, QVariant value)
{
    if(m_params.contains(key))
        return;

    m_params[key] = value;
    emit paramsChanged();
}

void QKlipperMessage::resetParams()
{
    m_params.clear();
    emit paramsChanged();
}

QString QKlipperMessage::method() const
{
    return m_method;
}

QString QKlipperMessage::toUri(bool parameterEncoded) const
{
    QString method = m_method;
    QString uri;

    //change the few that don't conform to a direct conversion
    if(method.contains("get_"))
        method.remove("get_");
    else if(method.contains("post_"))
        method.remove("post_");
    else if(method.contains("delete_"))
        method.remove("delete_");

    method.replace('.', '/');

    if(parameterEncoded && !m_params.isEmpty())
    {
        QStringList paramKeys = m_params.keys();
        QString params;

        foreach(QString key, paramKeys)
        {
            switch((QMetaType::Type)m_params[key].metaType().id())
            {
            case QMetaType::Bool:
                if(m_params[key].toBool())
                    params += QString("&%1=%2").arg(key).arg("true");
                else
                    params += QString("&%1=%2").arg(key).arg("false");
                break;

            case QMetaType::Int:
            case QMetaType::UInt:
            case QMetaType::LongLong:
            case QMetaType::ULongLong:
            case QMetaType::Long:
            case QMetaType::Short:
            case QMetaType::ULong:
            case QMetaType::UShort:
                params += QString("&%1=%2").arg(key).arg(m_params[key].toInt());
                break;
            case QMetaType::Double:
            case QMetaType::Float:
                params += QString("&%1=%2").arg(key).arg(m_params[key].toDouble());
                break;
            case QMetaType::Char:
            case QMetaType::Char16:
            case QMetaType::Char32:
            case QMetaType::UChar:
            case QMetaType::SChar:
            case QMetaType::QChar:
                break;
            case QMetaType::Nullptr:
            case QMetaType::Void:
            case QMetaType::VoidStar:
                params += QString("&%1").arg(key);
                break;
            case QMetaType::QString:
            case QMetaType::QByteArray:
                params += QString("&%1=%2").arg(key).arg(m_params[key].toString());
                break;
            case QMetaType::QStringList:
                break;
            case QMetaType::QBitArray:
            case QMetaType::QDateTime:
            case QMetaType::QDate:
            case QMetaType::QTime:
            case QMetaType::QUrl:
            case QMetaType::QLocale:
            case QMetaType::QRect:
            case QMetaType::QRectF:
            case QMetaType::QSize:
            case QMetaType::QSizeF:
            case QMetaType::QLine:
            case QMetaType::QLineF:
            case QMetaType::QPoint:
            case QMetaType::QPointF:
            case QMetaType::QEasingCurve:
            case QMetaType::QUuid:
            case QMetaType::QVariant:
            case QMetaType::QRegularExpression:
            case QMetaType::QJsonValue:
            case QMetaType::QJsonObject:
            case QMetaType::QJsonArray:
            case QMetaType::QJsonDocument:
            case QMetaType::QCborValue:
            case QMetaType::QCborArray:
            case QMetaType::QCborMap:
            case QMetaType::Float16:
            case QMetaType::QModelIndex:
            case QMetaType::QPersistentModelIndex:
            case QMetaType::QObjectStar:
            case QMetaType::QVariantMap:
            case QMetaType::QVariantList:
            case QMetaType::QVariantHash:
            case QMetaType::QVariantPair:
            case QMetaType::QByteArrayList:
            case QMetaType::QFont:
            case QMetaType::QPixmap:
            case QMetaType::QBrush:
            case QMetaType::QColor:
            case QMetaType::QPalette:
            case QMetaType::QIcon:
            case QMetaType::QImage:
            case QMetaType::QPolygon:
            case QMetaType::QRegion:
            case QMetaType::QBitmap:
            case QMetaType::QCursor:
            case QMetaType::QKeySequence:
            case QMetaType::QPen:
            case QMetaType::QTextLength:
            case QMetaType::QTextFormat:
            case QMetaType::QTransform:
            case QMetaType::QMatrix4x4:
            case QMetaType::QVector2D:
            case QMetaType::QVector3D:
            case QMetaType::QVector4D:
            case QMetaType::QQuaternion:
            case QMetaType::QPolygonF:
            case QMetaType::QColorSpace:
            case QMetaType::QSizePolicy:
            case QMetaType::UnknownType:
            case QMetaType::User:
            case QMetaType::QCborSimpleType:
                params += QString("&%1").arg(key);
                break;
            }
        }

        if(params.startsWith('&'))
            params.removeAt(0);

        uri = QString("/%1?%2").arg(method).arg(QUrl::toPercentEncoding(params, QByteArray("&=")));
    }
    else
        uri = QString("/%1").arg(method);

    return uri;
}

QByteArray QKlipperMessage::toRpc() const
{
    QJsonDocument document;
    QJsonObject rootObject;
    QJsonObject paramsObject;

    if(m_params.count() > 0)
    {
        QStringList paramKeys = m_params.keys();

        foreach(QString param, paramKeys)
        {
            switch((QMetaType::Type)m_params[param].metaType().id())
            {
            case QMetaType::Bool:
                paramsObject[param] = m_params[param].toBool();
                break;

            case QMetaType::Int:
            case QMetaType::UInt:
            case QMetaType::LongLong:
            case QMetaType::ULongLong:
            case QMetaType::Long:
            case QMetaType::Short:
            case QMetaType::ULong:
            case QMetaType::UShort:
                paramsObject[param] = m_params[param].toInt();
                break;
            case QMetaType::Double:
                paramsObject[param] = m_params[param].toDouble();
                break;
            case QMetaType::Float:
                paramsObject[param] = m_params[param].toFloat();
                break;
            case QMetaType::Char:
            case QMetaType::Char16:
            case QMetaType::Char32:
            case QMetaType::UChar:
            case QMetaType::SChar:
            case QMetaType::QChar:
                paramsObject[param] = m_params[param].toString();
                break;
            case QMetaType::QString:
            case QMetaType::QByteArray:
                paramsObject[param] = m_params[param].toString();
                break;
            case QMetaType::QStringList:
                break;
            case QMetaType::Nullptr:
            case QMetaType::Void:
            case QMetaType::VoidStar:
            case QMetaType::QBitArray:
            case QMetaType::QDateTime:
            case QMetaType::QDate:
            case QMetaType::QTime:
            case QMetaType::QUrl:
            case QMetaType::QLocale:
            case QMetaType::QRect:
            case QMetaType::QRectF:
            case QMetaType::QSize:
            case QMetaType::QSizeF:
            case QMetaType::QLine:
            case QMetaType::QLineF:
            case QMetaType::QPoint:
            case QMetaType::QPointF:
            case QMetaType::QEasingCurve:
            case QMetaType::QUuid:
            case QMetaType::QVariant:
            case QMetaType::QRegularExpression:
            case QMetaType::QJsonValue:
            case QMetaType::QJsonObject:
            case QMetaType::QJsonArray:
            case QMetaType::QJsonDocument:
            case QMetaType::QCborValue:
            case QMetaType::QCborArray:
            case QMetaType::QCborMap:
            case QMetaType::Float16:
            case QMetaType::QModelIndex:
            case QMetaType::QPersistentModelIndex:
            case QMetaType::QObjectStar:
            case QMetaType::QVariantMap:
            case QMetaType::QVariantList:
            case QMetaType::QVariantHash:
            case QMetaType::QVariantPair:
            case QMetaType::QByteArrayList:
            case QMetaType::QFont:
            case QMetaType::QPixmap:
            case QMetaType::QBrush:
            case QMetaType::QColor:
            case QMetaType::QPalette:
            case QMetaType::QIcon:
            case QMetaType::QImage:
            case QMetaType::QPolygon:
            case QMetaType::QRegion:
            case QMetaType::QBitmap:
            case QMetaType::QCursor:
            case QMetaType::QKeySequence:
            case QMetaType::QPen:
            case QMetaType::QTextLength:
            case QMetaType::QTextFormat:
            case QMetaType::QTransform:
            case QMetaType::QMatrix4x4:
            case QMetaType::QVector2D:
            case QMetaType::QVector3D:
            case QMetaType::QVector4D:
            case QMetaType::QQuaternion:
            case QMetaType::QPolygonF:
            case QMetaType::QColorSpace:
            case QMetaType::QSizePolicy:
            case QMetaType::UnknownType:
            case QMetaType::User:
            case QMetaType::QCborSimpleType:
                paramsObject[param];
                break;
            }
        }

        rootObject["params"] = paramsObject;
    }

    rootObject["method"] = m_method;
    rootObject["id"] = m_id;

    document.setObject(rootObject);

    QByteArray data = document.toJson(QJsonDocument::Compact) + QKLIPPER_RPC_EOF;

    return data;
}

void QKlipperMessage::setMethod(const QString &method)
{
    if (m_method == method)
        return;

    m_method = method;
    emit methodChanged();
}

void QKlipperMessage::resetMethod()
{
    setMethod(QString());
}

QKlipperMessage::State QKlipperMessage::state() const
{
    return m_state;
}

void QKlipperMessage::setState(State state)
{
    if (m_state == state)
        return;

    m_state = state;
    emit stateChanged();
}

void QKlipperMessage::resetState()
{
    setState(None);
}

QJsonValue QKlipperMessage::response() const
{
    return m_response;
}

void QKlipperMessage::setResponse(const QJsonValue &response)
{
    if(m_responseTimer->isActive())
        m_responseTimer->stop();

    m_response = response;

    emit responseChanged();
}

void QKlipperMessage::setResponse(const QByteArray &response)
{
    if(m_responseTimer->isActive())
        m_responseTimer->stop();

    QJsonParseError parseError;
    QJsonDocument document = QJsonDocument::fromJson(response, &parseError);

    if(parseError.error != QJsonParseError::NoError)
    {
        setState(Error);
        setErrorString(parseError.errorString());

        return;
    }

    setResponse(document.object());
}

void QKlipperMessage::resetResponse()
{
    setResponse(QJsonObject());
}

QString QKlipperMessage::errorString() const
{
    return m_errorString;
}

void QKlipperMessage::setErrorString(const QString &errorString)
{
    if (m_errorString == errorString)
        return;

    m_errorString = errorString;
    emit errorStringChanged();
}

void QKlipperMessage::resetErrorString()
{
    setErrorString(QString());
}

void QKlipperMessage::startTimer()
{
    if(m_responseTimer->isActive())
        m_responseTimer->stop();

    m_responseTimer->start();
}

void QKlipperMessage::on_responseTimerTimeout()
{
    emit responseTimeout();
}

QKlipperMessage::Origin QKlipperMessage::origin() const
{
    return m_origin;
}

void QKlipperMessage::setOrigin(Origin origin)
{
    if (m_origin == origin)
        return;

    m_origin = origin;
    emit originChanged();
}

void QKlipperMessage::resetOrigin()
{
    setOrigin(System);
}
